#263: Refactor Tugboat integration

rabbitlair opened the pull request:
Closes #263 

**Warning**: Contains a breaking change, so it requires, at least, a minor version bump.

This PR refactors the Tugboat integration to use DDEV configuration as the single source of truth for all service definitions. The new implementation generates a complete `config.yml` using Twig templates instead of separate bash scripts.

### Architecture
- **New dedicated plugin**: `TugboatConfigPlugin` handles all Tugboat logic (previously in `ScaffoldInstallerPlugin`)
- **Template-based generation**: Uses Twig template to generate `config.yml` file with sub-templates for each build phase in main PHP service
- **DDEV-driven configuration**: All service versions and types extracted from `.ddev/config.yaml` and docker-compose add-on files

### Build Phase Templates
- `config.yml.twig` - Main configuration template that includes phase templates
- `php-init.yml.twig` - System packages, PHP extensions, Node.js installation, composer install
- `php-update.yml.twig` - Composer install, sync task, file permissions changes, update tasks
- `php-build.yml.twig` - Build tasks

### Taskfile Integration
- **All services**: Supports `tugboat:{service}:init` tasks (e.g., `tugboat:redis:init`)
- **PHP service**: Supports `sync:tugboat`, `update:tugboat`, `build:tugboat`, `online:tugboat` tasks
- **Custom init script**: `custom-init-command.sh` wrapper for non-PHP service init tasks
- **Task installation**: Automatically installs Taskfile in init phase using version from `.taskfile` file

### Template Customization
- **Default templates**: Located in `vendor/lullabot/drainpipe/scaffold/tugboat/templates/`
- **Project overrides**: Place custom templates in `.tugboat/drainpipe-templates/`
- **Template precedence**: Project templates take priority over defaults
- **Custom Twig filters**: Added `indent` filter for proper YAML formatting

### Files Structure

```
scaffold/tugboat/
├── templates/
│   ├── config.yml.twig           # Main Tugboat config
│   ├── php-init.yml.twig         # Init phase commands
│   ├── php-update.yml.twig       # Update phase commands
│   ├── php-build.yml.twig        # Build phase commands
│   └── settings.tugboat.php.twig # Drupal settings
└── scripts/
    ├── custom-init-command.sh    # Init wrapper for non-PHP services
    └── install-mysql-client.sh   # MySQL client installer

.tugboat/ (generated)
├── config.yml                    # Generated config (overwritten on each run)
├── config.drainpipe-override.yml # User overrides (preserved)
├── drainpipe-templates/          # User template overrides (preserved)
└── scripts/                      # Copied scripts
```

Please, note this is a first working version. I would like to know your thoughts about the new approach - any feedback will be more than welcome. As far as I can tell, this refactor does not break backwards compatibility while adding the custom templates feature, which I think will make easier for users to customize their Tugboat integration. 

## Testing Instructions

First, create a new Drupal project and configure Drainpipe with Tugboat integration.

```
# Create a new Drupal project
mkdir testing
cd testing
composer create-project drupal/recommended-project . --ignore-platform-reqs
# Copy this Drainpipe branch to `testing/drainpipe` 

# Configure DDEV
ddev config --auto
ddev config --php-version "8.4"
ddev config --nodejs-version "24"
ddev get ddev/ddev-redis --version v1.2.0
ddev get ddev/ddev-elasticsearch --version v0.3.2
ddev start

# Configure Composer to work with Drainpipe
ddev exec --raw composer config extra.drupal-scaffold.gitignore true
ddev exec --raw composer config --json extra.drupal-scaffold.allowed-packages \[\"lullabot/drainpipe\"]
ddev exec --raw composer config --no-plugins allow-plugins.composer/installers true
ddev exec --raw composer config --no-plugins allow-plugins.drupal/core-composer-scaffold true
ddev exec --raw composer config --no-plugins allow-plugins.lullabot/drainpipe true
ddev exec --raw composer config repositories.drainpipe --json '{"type": "path", "url": "drainpipe", "options": {"symlink": false}}'
ddev exec --raw composer config minimum-stability dev

# Configure Tugboat integration
ddev exec --raw composer config extra.drainpipe --json '{"tugboat": {}}'
mkdir -p .tugboat
echo "php:" >  .tugboat/config.drainpipe-override.yml
echo "  aliases:" >> .tugboat/config.drainpipe-override.yml
echo "    - foo" >> .tugboat/config.drainpipe-override.yml
echo "  urls:" >> .tugboat/config.drainpipe-override.yml
echo "    - /" >> .tugboat/config.drainpipe-override.yml
echo "    - /?v=1" >> .tugboat/config.drainpipe-override.yml
echo "  screenshot:" >> .tugboat/config.drainpipe-override.yml
echo "    timeout: 45" >> .tugboat/config.drainpipe-override.yml
echo "  visualdiff:" >> .tugboat/config.drainpipe-override.yml
echo "    fullPage: false" >> .tugboat/config.drainpipe-override.yml

# Install Drainpipe
ddev composer require lullabot/drainpipe --with-all-dependencies
```

Verify the generated `.tugboat/config.yml` file is correct, and the services declared match the DDEV settings.

#### Test scenarios
1. Override a template for a PHP service's build phase
```
mkdir -p .tugboat/drainpipe-templates
cp vendor/lullabot/drainpipe/scaffold/tugboat/templates/php-update.yml.twig .tugboat/drainpipe-templates
# Add some new command to the file
ddev composer install
```
Then verify the new command is included in `.tugboat/config.yml`.

2. Add a custom `init` command for a secondary service:
```
echo "" >> Taskfile.yml
echo "  tugboat:redis:init:" >> Taskfile.yml
echo "    cmds:" >> Taskfile.yml
echo "      - whoami" >> Taskfile.yml
ddev composer install
```
Then verify the new command is included in `.tugboat/config.yml`.

3. Pantheon integration:
```
ddev exec --raw composer config extra.drainpipe.tugboat --json '{"pantheon": true}'
ddev composer install
```
Then verify Pantheon integration is included in `.tugboat/config.yml`.

#### More testing

Tested in a project currently using Drainpipe, so we can be confident about nothing breaking. [Tested on lullabotdotcom.](https://github.com/Lullabot/drainpipe/pull/1073#issuecomment-3725191115)

deviantintegral commented on scaffold/tugboat/templates/config.yml.twig:
```diff
@@ -0,0 +1,73 @@
+#   environment:
+services:
+  php:
+    http: false
```
huh, should this be true?

rabbitlair commented:
@deviantintegral, Gemini comments were already addressed :)

Regarding test steps, that is a great idea. Let me create those, so this one is easier to review. Thanks!

rabbitlair commented:
I tested this PR in latest `main` from Lullabot D8 project, and it works like a charm. The only detail which is different is the calls in the build step are different. Currently, these are executed:
```
task build:tugboat
task drupal:update
task tugboat:drush-uli-ready
```
while the new branch replaces those to be:
```
task build
task update
task build:tugboat   # This is equivalent to tugboat:drush-uli-ready
```
We would need to update Lullabot-D8 to override the php-update Twig template, but apart from that... it should be good!

YesCT commented:
There's a note in the description:

> Also, take into account tests are broken - once we all are happy with the approach and implementation, I will fix those tests failing.

But I see 3 recent commits with passing tests green checkmarks. Hm. 

Are tests expected to be broken? Maybe that comment in the description is outdated. Or If it is still accurate, maybe it could be updated to list which tests. It's probably outdated. I'm gonna remove it. Put it back if I guessed wrong. ;)

YesCT commented on .github/workflows/TestTugboat.yml:
```diff
@@ -119,7 +110,6 @@ jobs:
         run: |
           ddev config --auto
           ddev config --php-version "8.3"
-          ddev config --nodejs-version "16"
```
[Non-blocking Questions] 

why take this out?

cause node 16 is sooooo old we should be testing it?? ha :D 

do we test node override elsewhere?

why leave the php version config in??

hm.

YesCT commented on .tugboat/config.yml:
```diff
@@ -19,17 +20,89 @@ services:
-
+      init:
+        - |
+          apt-get update
```
hm. we loose the ability to run an automatic shell script syntax checker. 
why not keep the 1-init.sh pattern??

YesCT commented on scaffold/tugboat/scripts/install-mysql-client.sh:
[Non-blocking Question] Is this an intended change?

If so, as part of review, should we check all the scripts (in scaffold/tugboat/scripts ??)are 644??

rabbitlair replied:
Yes, it is an intended change - this script, and the other implemented in this refactor, are both scaffolded by Drainpipe, so I felt it was a good practice to remove the write permissions from them. I am open to keep the write permissions if we think that is better, of course

YesCT commented on scaffold/tugboat/steps/1-init.sh.twig:
Just noting, we were doing twig templates before, this PR isn't introducing that.

YesCT commented on src/TugboatConfigPlugin.php:
```diff
@@ -0,0 +1,588 @@
+            return $matches[2];
+        }
+
+        // Extract version from image:version format
```
Thanks for the example format note.

YesCT commented on src/TugboatConfigPlugin.php:
```diff
@@ -0,0 +1,588 @@
+        if (preg_match('/^\$\{[^:}]+:-(.+)\}$/', $image, $matches)) {
+            $image = $matches[1];
+        }
+        // Example: redis:${REDIS_TAG:-7-alpine}
```
<3

YesCT commented on README.md:
```diff
@@ -880,11 +923,8 @@ php:
   checkout:
   depends:
-  aliases:
-  urls:
```
Are these changes unrelated to the refactor? Maybe removing invalid options that were invalid before even, could be it's own PR?? If so... [Request for changes]

rabbitlair commented on .tugboat/config.yml:
```diff
@@ -19,17 +20,89 @@ services:
+      init:
+        - |
+          apt-get update
+          apt-get install -y mariadb-client
```
I tried to preserve all logic in the original shell scripts, but will double check in case I have missed something

rabbitlair commented on scaffold/tugboat/templates/settings.tugboat.php.twig:
```diff
@@ -16,7 +16,7 @@ if (getenv('TUGBOAT_REPO')) {
     'password' => 'tugboat',
     'prefix' => '',
-    'host' => '{{ database_type }}',
+    'host' => '{{ services.database.type }}',
```
This change corresponds to the change on how the Twig variables are managed after the refactor. Instead of providing individual variables, everything related to Tugboat services is now passed as a single `services` array, hence the change from `database_type` to `services.database.type`.

YesCT commented on .tugboat/config.yml:
[Very non-blocking] Would https://github.com/Lullabot/drainpipe/issues/771 have covered this?

rabbitlair commented:
After testing this branch with another real world project that has a complex Tugboat integration, I added the following changes:
* New template `php-online.yml.twig` to easily customize Online build phase
* Fixed a bug that prevented database service to properly run its custom init commands
* Fixed a bug when installing Taskfile to run custom init commands for all services
* Updated the MySQL repository GPG key, as the current one is expired
* Implemented support for using Percona as the Docker image for database service
